docs/CONFIGURATION.md
# Configuration

You can generate default config, using `relay --gen-config 'config.yaml'`

## Example config with graphql transport

```yaml
---
listen_address: "127.0.0.1:12345"
keys_path: /var/lib/relay/keys.json
storage_path: /var/lib/relay/persistent_storage
logger_settings:
  appenders:
    stdout:
      kind: console
      encoder:
        pattern: "{d(%Y-%m-%d %H:%M:%S %Z)(utc)} - {h({l})} {M} {f}:{L} = {m} {n}"
  root:
    level: error
    appenders:
      - stdout
  loggers:
    relay:
      level: info
      appenders:
        - stdout
      additive: false
    relay_eth:
      level: info
      appenders:
        - stdout
      additive: false
    relay_ton:
      level: info
      appenders:
        - stdout
      additive: false
metrics_settings:
  listen_address: "127.0.0.1:10000"
  metrics_path: "/"
  collection_interval: 10s
eth_settings:
  node_address: "http://localhost:8545"
  tcp_connection_count: 100
  get_eth_data_timeout: 10s
  get_eth_data_attempts: 50
  eth_poll_interval: 10s
  eth_poll_attempts: 8640
  suspicious_blocks_offset: 100
  bridge_address: ""
ton_settings:
  bridge_contract_address: ""
  relay_contract_address: ""
  transport:
    type: graphql
    address: "https://main.ton.dev/graphql"
    next_block_timeout: 60s
    parallel_connections: 100
    fetch_timeout: 10s
    retry_delay: 500ms
  event_configuration_details_retry_interval: 5s
  event_configuration_details_retry_count: 100
  event_details_retry_interval: 5s
  event_details_retry_count: 100
  message_retry_interval: 60s
  message_retry_count: 10
  message_retry_interval_multiplier: 1.5
  parallel_spawned_contracts_limit: 10
  ton_events_verification_interval: 1s
  ton_events_verification_queue_lt_offset: 10
  ton_events_allowed_time_diff: 10
  events_handler_retry_count: 50
  events_handler_interval: 10s
``` 

- `keys_path` path to file, where encrypted data is stored.
- `storage_path` path for database
- `listen_address` address to bind control server.  **EXPOSING IT TO OUTER WORLD
  IS PROHIBITED**, because anyone, having access to it can control relay.
- `number_of_ethereum_tcp_connections` maximum number of parallel tcp
  connections to ethereum node

### metrics_settings

- `listen_address` - address to bind metrics server
- `metrics_path` - the HTTP resource path on which to fetch metrics from
  targets.
- `collection_interval` - metrics poll interval. Requests between collections
  will return cached data

### eth_settings

- `node_address`  - address of ethereum node
- `tcp_connection_count` - maximum number of parallel tcp connections to
  ethereum node
- `get_eth_data_timeout` - timeout and delay between retries for getting
  non-critical data, like current eth sync status or current height
- `get_eth_data_attempts` - number of attempts for previous field
- `eth_poll_interval` - poll interval between fetching new blocks
- `eth_poll_attempts`  - number of attempts to get logs in the block
- `suspicious_blocks_offset` - offset in blocks for checking suspicious
  transactions
- `bridge_address` - address of bridge contract in ethereum

### ton_settings

- `bridge_contract_address` - address of bridge contract
- `relay_contract_address` - address of relay contract

  *Next section is optional to configure, default settings are just ok*

- `event_configuration_details_retry_interval` - time to wait between retries
- `event_configuration_details_retry_count` - times to get configuration details
  in case of errors
- `event_details_retry_interval` - time to wait between retries
- `event_details_retry_count` - times to get event details in case of errors
- `message_retry_interval` - interval between resending messages
- `message_retry_count`- number of retries to send message
- `message_retry_interval_multiplier`- coefficient, on which every interval will
  be multiplied
- `parallel_spawned_contracts_limit`- amount of parallel sent messages
- `ton_events_verification_interval`- interval for verification queue processing
  loop
- `ton_events_verification_queue_lt_offset` - lt delay before current logical
  time

#### GraphQL

- `address` - address of graphql endpoint
- `next_block_timeout`  - timeout for blocks emission
- `fetch_timeout` - timeout for GraphQL queries
- `parallel_connections` - amount of parallel connections to GraphQL

### Tonlib

- `server_address` address of ton lite server
- `server_key` key of lite server
- `last_block_threshold`  last block id caching duration
- `subscription_polling_interval` how often accounts are polled. Has sense when
  it's greater or equal `last_block_threshold_sec`


## How to use

### First run

We don't provide prebuilt `.deb` packages due security reasons. So, to
get `.deb`
you should run this:

- `./scripts/build_deb.sh`
- run `sudo dpkg -i name_of_deb_package_you_got_on_previous_step`
- change config in you favourite editor (default location is
  `/etc/relay.conf`)
- run it: `sudo systemctl start relay`
- init it: ` relay-client --server-addr ADDRESS_YOU_SET_AS_LISTEN_ADDRESS` and
  enjoy cli experience.

### Service restart

- run client and unlock the relay

### Db operations

If you want to fully reinitialize relay for some reason, remove all files
in `keys_path` and in `storage_path`


src/config.rs
use std::net::SocketAddr;
use std::path::{Path, PathBuf};
use std::time::Duration;

use anyhow::Error;
use clap::Clap;
use config::{Config, File, FileFormat};
use http::uri::PathAndQuery;

#[cfg(feature = "graphql-transport")]
use relay_ton::transport::graphql_transport::Config as TonGraphQLConfig;
#[cfg(feature = "tonlib-transport")]
use relay_ton::transport::tonlib_transport::Config as TonTonlibConfig;

use crate::prelude::*;

#[derive(Deserialize, Serialize, Clone, Debug, Default, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct TonAddress(pub String);

#[derive(Deserialize, Serialize, Clone, Debug)]
pub struct RelayConfig {
    /// Listen address of relay. Used by the client to perform all maintenance actions.
    pub listen_address: SocketAddr,

    /// Path to json, where ton and eth private keys will be stored in encrypted way.
    pub keys_path: PathBuf,

    /// Path to Sled database.
    pub storage_path: PathBuf,

    /// Logger settings
    #[serde(default = "default_logger_settings")]
    pub logger_settings: serde_yaml::Value,

    /// Metrics settings
    #[serde(default)]
    pub metrics_settings: Option<MetricsSettings>,

    /// ETH specific settings
    pub eth_settings: EthSettings,

    /// TON specific settings
    pub ton_settings: TonSettings,
}

#[derive(Deserialize, Serialize, Clone, Debug)]
pub struct MetricsSettings {
    /// Listen address of metrics. Used by the client to gather prometheus metrics
    pub listen_address: SocketAddr,

    /// Listen address of metrics. Used by the client to gather prometheus metrics
    #[serde(default = "default_metrics_path", with = "relay_utils::serde_url")]
    pub metrics_path: PathAndQuery,

    /// Metrics poll interval
    #[serde(with = "relay_utils::serde_time")]
    pub collection_interval: Duration,
}

#[derive(Deserialize, Serialize, Clone, Debug)]
pub struct EthSettings {
    /// Address of ethereum node. Only http is supported right now
    pub node_address: String,

    /// Number of concurrent tcp connection to ethereum node
    pub tcp_connection_count: usize,

    /// Timeout and delay between retries for getting non-critical data, like current eth sync status
    #[serde(with = "relay_utils::serde_time")]
    pub get_eth_data_timeout: Duration,

    /// Number of attempts  for getting non-critical data, like current eth sync status
    pub get_eth_data_attempts: u64,

    /// Poll interval between fetching new blocks
    #[serde(with = "relay_utils::serde_time")]
    pub eth_poll_interval: Duration,

    /// Number of attempts to get logs in the block
    pub eth_poll_attempts: u64,

    /// Offset in blocks for checking suspicious transactions
    pub suspicious_blocks_offset: u64,

    /// Bridge address in ethereum
    pub bridge_address: Address,
}

impl Default for EthSettings {
    fn default() -> Self {
        Self {
            node_address: "http://localhost:8545".into(),
            tcp_connection_count: 100,
            get_eth_data_timeout: Duration::from_secs(10),
            get_eth_data_attempts: 50,
            eth_poll_interval: Duration::from_secs(10),
            eth_poll_attempts: 86400 / 10,
            suspicious_blocks_offset: 10,
            bridge_address: Default::default(),
        }
    }
}

#[derive(Deserialize, Serialize, Clone, Debug)]
pub struct TonSettings {
    /// Relay account address
    pub relay_contract_address: TonAddress,

    /// Bridge contract address
    pub bridge_contract_address: TonAddress,

    /// TON transport config
    pub transport: TonTransportConfig,

    /// Interval between attempts to get event configuration details
    #[serde(with = "relay_utils::serde_time")]
    pub event_configuration_details_retry_interval: Duration,
    /// Amount of unsuccessful attempts
    pub event_configuration_details_retry_count: u64,

    /// Interval between attempts to get event details
    #[serde(with = "relay_utils::serde_time")]
    pub event_details_retry_interval: Duration,
    /// Amount of unsuccessful attempts
    pub event_details_retry_count: u64,

    /// Interval between attempts to get send message
    #[serde(with = "relay_utils::serde_time")]
    pub message_retry_interval: Duration,
    /// Amount of unsuccessful attempts
    pub message_retry_count: u64,
    /// Coefficient, on which every interval will be multiplied
    pub message_retry_interval_multiplier: f64,

    /// Amount of parallel sent messages in ton
    pub parallel_spawned_contracts_limit: usize,

    /// TON events verification interval
    #[serde(with = "relay_utils::serde_time")]
    pub ton_events_verification_interval: Duration,

    /// TON events verification queue logical time offset
    pub ton_events_verification_queue_lt_offset: u64,
    /// TON events allowed timestamp difference
    pub ton_events_allowed_time_diff: u32,

    /// Amount of attempts to create events handler
    pub events_handler_retry_count: u64,
    /// Duration between attempts to create events handler
    #[serde(with = "relay_utils::serde_time")]
    pub events_handler_interval: Duration,
}

impl Default for TonSettings {
    fn default() -> Self {
        Self {
            relay_contract_address: Default::default(),
            bridge_contract_address: Default::default(),
            transport: TonTransportConfig::default(),
            event_configuration_details_retry_count: 100,
            event_configuration_details_retry_interval: Duration::from_secs(5),
            event_details_retry_interval: Default::default(),
            event_details_retry_count: 100,
            message_retry_interval: Duration::from_secs(60),
            message_retry_count: 10,
            message_retry_interval_multiplier: 1.5,
            parallel_spawned_contracts_limit: 10,
            ton_events_verification_interval: Duration::from_secs(1),
            ton_events_verification_queue_lt_offset: 10,
            ton_events_allowed_time_diff: 10,
            events_handler_retry_count: 50,
            events_handler_interval: Duration::from_secs(10),
        }
    }
}

impl Default for RelayConfig {
    fn default() -> Self {
        Self {
            keys_path: PathBuf::from("/var/lib/relay/keys.json"),
            listen_address: "127.0.0.1:12345".parse().unwrap(),
            storage_path: PathBuf::from("/var/lib/relay/persistent_storage"),
            logger_settings: default_logger_settings(),
            metrics_settings: Some(MetricsSettings {
                listen_address: "127.0.0.1:10000".parse().unwrap(),
                metrics_path: default_metrics_path(),
                collection_interval: Duration::from_secs(10),
            }),
            eth_settings: EthSettings::default(),
            ton_settings: TonSettings::default(),
        }
    }
}

fn default_logger_settings() -> serde_yaml::Value {
    const DEFAULT_LOG4RS_SETTINGS: &str = r##"
    appenders:
      stdout:
        kind: console
        encoder:
          pattern: "{d(%Y-%m-%d %H:%M:%S %Z)(utc)} - {h({l})} {M} {f}:{L} = {m} {n}"
    root:
      level: error
      appenders:
        - stdout
    loggers:
      relay:
        level: info
        appenders:
          - stdout
        additive: false
      relay_eth:
        level: info
        appenders:
          - stdout
        additive: false
      relay_ton:
        level: info
        appenders:
          - stdout
        additive: false
    "##;
    serde_yaml::from_str(DEFAULT_LOG4RS_SETTINGS).unwrap()
}

fn default_metrics_path() -> PathAndQuery {
    PathAndQuery::from_static("/")
}

#[derive(Deserialize, Serialize, Clone, Debug)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum TonTransportConfig {
    #[cfg(feature = "tonlib-transport")]
    Tonlib(TonTonlibConfig),
    #[cfg(feature = "graphql-transport")]
    GraphQl(TonGraphQLConfig),
}

#[cfg(any(feature = "tonlib-transport", feature = "graphql-transport"))]
impl Default for TonTransportConfig {
    fn default() -> Self {
        #[cfg(feature = "tonlib-transport")]
        return Self::Tonlib(TonTonlibConfig::default());

        #[cfg(all(feature = "graphql-transport", not(feature = "tonlib-transport")))]
        Self::GraphQl(TonGraphQLConfig::default())
    }
}

pub fn read_config(path: PathBuf) -> Result<RelayConfig, Error> {
    let mut config = Config::new();
    config.merge(File::from(path).format(FileFormat::Yaml))?;
    config.merge(config::Environment::new())?;

    let config: RelayConfig = config.try_into()?;
    Ok(config)
}

pub fn read_env() -> Result<RelayConfig, Error> {
    let config_string = std::env::var("config_string")
        .map_err(|e| anyhow::Error::new(e).context("config_string not set"))?;
    let config = serde_json::from_str(&config_string)?;
    Ok(config)
}

#[derive(Deserialize, Serialize, Clone, Debug, Clap)]
pub struct Arguments {
    /// Path to config
    #[clap(short, long, conflicts_with = "gen-config")]
    pub config: Option<PathBuf>,

    /// Generate default config
    #[clap(long)]
    pub gen_config: Option<PathBuf>,
}

pub fn generate_config<T>(path: T) -> Result<(), Error>
where
    T: AsRef<Path>,
{
    let mut file = std::fs::File::create(path)?;
    let config = RelayConfig::default();
    file.write_all(serde_yaml::to_string(&config)?.as_bytes())?;
    Ok(())
}

pub fn parse_args() -> Arguments {
    Arguments::parse()
}
